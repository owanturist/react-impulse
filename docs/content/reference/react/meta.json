{
  "pages": [
    "use-scope",
    "use-scoped",
    "use-scoped-memo",
    "use-scoped-callback",
    "use-scoped-effect",
    "use-scoped-layout-effect"
  ]
}
