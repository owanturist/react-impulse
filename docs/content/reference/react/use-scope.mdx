---
title: useScope
description: API documentation for the useScope React hook
---

## Overview

The `useScope` hook returns a reactive scope tied to the component lifecycle.

## Function Signature

```typescript
function useScope(): Scope
```

## Returns

A `Scope` instance that triggers component re-renders when read impulses change.

## Description

`useScope` provides a scope that enables reactive impulse reads within React components. When impulses are read within this scope, the component automatically re-renders when those impulses change.

## Examples

```typescript
function Counter() {
  const scope = useScope()
  const impulse = Impulse(0)

  return (
    <div>
      Count: {impulse.getValue(scope)}
    </div>
  )
}
```