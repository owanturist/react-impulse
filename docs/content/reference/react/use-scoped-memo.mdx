---
title: useScopedMemo
description: API documentation for the useScopedMemo React hook
---

## Overview

The `useScopedMemo` hook memoizes computations that depend on impulses, similar to `React.useMemo` but with reactive scope support.

## Function Signature

```typescript
function useScopedMemo<T>(
  factory: (scope: Scope) => T,
  dependencies: DependencyList,
): T
```

## Parameters

- `factory`: A function that receives a `Scope` and computes the memoized value
- `dependencies`: An array of dependencies that control when the memoization is invalidated

## Returns

The memoized value of type `T`.

## Description

`useScopedMemo` combines impulse reactivity with React's memoization. The factory function is re-executed when either the dependencies change or when any impulses read within the scope change.

## Examples

```typescript
const memoizedValue = useScopedMemo(
  (scope) => expensiveComputation(impulse.getValue(scope)),
  [someDependency],
)
```
