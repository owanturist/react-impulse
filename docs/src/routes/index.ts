import { createFileRoute } from "@tanstack/react-router"

import { Home } from "~/pages/home"

export const Route = createFileRoute("/")({
  component: Home,
})
