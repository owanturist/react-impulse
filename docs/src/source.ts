export { create, docs as source } from "../source.generated"
