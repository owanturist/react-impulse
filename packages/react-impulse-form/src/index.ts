export * from "./impulse-form"

export * from "./impulse-form-unit"
export * from "./impulse-form-shape"
export type * from "./impulse-form-meta"
export * from "./impulse-form-list"
export * from "./impulse-form-switch"
export * from "./impulse-form-optional"

export type * from "./result"
export type * from "./zod-like-schema"
export type { ValidateStrategy } from "./validate-strategy"
