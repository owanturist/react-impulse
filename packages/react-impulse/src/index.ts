export type { DependencyList } from "./dependencies"
export type { Compare, Destructor } from "./utils"
export {
  type ImpulseOptions,
  type TransmittingImpulseOptions,
  type ReadonlyImpulse,
  type ImpulseGetter,
  Impulse,
} from "./Impulse"
export type { Scope } from "./Scope"
export { untrack } from "./untrack"
export { batch } from "./batch"
export { batch as tap } from "./batch"
export { subscribe } from "./subscribe"
export { useScope } from "./useScope"
export { type UseScopedOptions, useScoped } from "./useScoped"
export { useScopedMemo } from "./useScopedMemo"
export { useScopedCallback } from "./useScopedCallback"
export { useScopedEffect } from "./useScopedEffect"
export { useScopedLayoutEffect } from "./useScopedLayoutEffect"
