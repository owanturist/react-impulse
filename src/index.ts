export type { Dispatch } from "react"
export type { Compare, SetInnerState } from "./utils"
export type { ExtractInnerState, DeepExtractInnerState } from "./InnerStore"
export { batch } from "./batch"
export { InnerStore } from "./InnerStore"
export { useGetInnerState } from "./useGetInnerState"
export { useSetInnerState } from "./useSetInnerState"
export { useInnerState } from "./useInnerState"
export { useInnerReducer } from "./useInnerReducer"
export { useInnerWatch } from "./useInnerWatch"
