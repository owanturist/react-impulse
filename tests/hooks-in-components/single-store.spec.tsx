import React from "react"
import { act, render, screen, fireEvent } from "@testing-library/react"

import { Impulse, useSweetyState } from "../../src"
import { Counter } from "../common"

import { withinNth } from "./common"

describe("single store", () => {
  const GetterComponent: React.FC<{
    store: Impulse<Counter>
    onRender: VoidFunction
  }> = ({ store, onRender }) => {
    const state = useSweetyState(store)

    return (
      <React.Profiler id="test" onRender={onRender}>
        <span data-testid="getter">{state.count}</span>
      </React.Profiler>
    )
  }

  const SetterComponent: React.FC<{
    store: Impulse<Counter>
    onRender: VoidFunction
  }> = ({ store, onRender }) => (
    <React.Profiler id="test" onRender={onRender}>
      <div data-testid="setter">
        <button
          type="button"
          data-testid="increment"
          onClick={() => store.setState(Counter.inc)}
        />
        <button
          type="button"
          data-testid="reset"
          onClick={() => store.setState({ count: 0 }, Counter.compare)}
        />
      </div>
    </React.Profiler>
  )

  const SingleSetterSingleGetter: React.FC<{
    store: Impulse<Counter>
    onRootRender: VoidFunction
    onGetterRender: VoidFunction
    onSetterRender: VoidFunction
  }> = ({ store, onRootRender, onGetterRender, onSetterRender }) => (
    <>
      <React.Profiler id="test" onRender={onRootRender} />

      <GetterComponent store={store} onRender={onGetterRender} />
      <SetterComponent store={store} onRender={onSetterRender} />
    </>
  )

  it("Single Setter / Getter", () => {
    const store = Impulse.of({ count: 0 })
    const onRootRender = vi.fn()
    const onGetterRender = vi.fn()
    const onSetterRender = vi.fn()

    render(
      <SingleSetterSingleGetter
        store={store}
        onRootRender={onRootRender}
        onGetterRender={onGetterRender}
        onSetterRender={onSetterRender}
      />,
    )

    // check initial state
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(1)
    expect(screen.getByTestId("getter")).toHaveTextContent("0")

    // increment by from the component
    fireEvent.click(screen.getByTestId("increment"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(2)
    expect(screen.getByTestId("getter")).toHaveTextContent("1")

    // increment from the outside
    act(() => store.setState(Counter.inc))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(3)
    expect(screen.getByTestId("getter")).toHaveTextContent("2")

    // reset from the component
    fireEvent.click(screen.getByTestId("reset"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(4)
    expect(screen.getByTestId("getter")).toHaveTextContent("0")

    // reset second time in a row
    fireEvent.click(screen.getByTestId("reset"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(4)
    expect(screen.getByTestId("getter")).toHaveTextContent("0")

    // increment twice in a row
    fireEvent.click(screen.getByTestId("increment"))
    fireEvent.click(screen.getByTestId("increment"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(6)
    expect(screen.getByTestId("getter")).toHaveTextContent("2")

    // increment twice in a row from the outside
    act(() => {
      store.setState(Counter.inc)
      store.setState(Counter.inc)
    })
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onSetterRender).toHaveBeenCalledTimes(1)
    expect(onGetterRender).toHaveBeenCalledTimes(7)
    expect(screen.getByTestId("getter")).toHaveTextContent("4")
  })

  const MultipleSetterMultipleGetter: React.FC<{
    store: Impulse<Counter>
    onRootRender: VoidFunction
    onFirstGetterRender: VoidFunction
    onSecondGetterRender: VoidFunction
    onFirstSetterRender: VoidFunction
    onSecondSetterRender: VoidFunction
  }> = ({
    store,
    onRootRender,
    onFirstGetterRender,
    onSecondGetterRender,
    onFirstSetterRender,
    onSecondSetterRender,
  }) => (
    <>
      <React.Profiler id="test" onRender={onRootRender} />
      <GetterComponent store={store} onRender={onFirstGetterRender} />
      <GetterComponent store={store} onRender={onSecondGetterRender} />
      <SetterComponent store={store} onRender={onFirstSetterRender} />
      <SetterComponent store={store} onRender={onSecondSetterRender} />
    </>
  )

  it("Multiple Setters / Getters", () => {
    const store = Impulse.of({ count: 0 })
    const onRootRender = vi.fn()
    const onFirstGetterRender = vi.fn()
    const onSecondGetterRender = vi.fn()
    const onFirstSetterRender = vi.fn()
    const onSecondSetterRender = vi.fn()

    render(
      <MultipleSetterMultipleGetter
        store={store}
        onRootRender={onRootRender}
        onFirstGetterRender={onFirstGetterRender}
        onSecondGetterRender={onSecondGetterRender}
        onFirstSetterRender={onFirstSetterRender}
        onSecondSetterRender={onSecondSetterRender}
      />,
    )

    // check initial state
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onFirstSetterRender).toHaveBeenCalledTimes(1)
    expect(onSecondSetterRender).toHaveBeenCalledTimes(1)
    expect(onFirstGetterRender).toHaveBeenCalledTimes(1)
    expect(onSecondGetterRender).toHaveBeenCalledTimes(1)
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()

    // increment from the first component
    fireEvent.click(withinNth("setter", 0).getByTestId("increment"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onFirstSetterRender).toHaveBeenCalledTimes(1)
    expect(onSecondSetterRender).toHaveBeenCalledTimes(1)
    expect(onFirstGetterRender).toHaveBeenCalledTimes(2)
    expect(onSecondGetterRender).toHaveBeenCalledTimes(2)
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()

    // increment from the second component
    fireEvent.click(withinNth("setter", 1).getByTestId("increment"))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onFirstSetterRender).toHaveBeenCalledTimes(1)
    expect(onSecondSetterRender).toHaveBeenCalledTimes(1)
    expect(onFirstGetterRender).toHaveBeenCalledTimes(3)
    expect(onSecondGetterRender).toHaveBeenCalledTimes(3)
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()

    // increment from the outside
    act(() => store.setState(Counter.inc))
    expect(onRootRender).toHaveBeenCalledTimes(1)
    expect(onFirstSetterRender).toHaveBeenCalledTimes(1)
    expect(onSecondSetterRender).toHaveBeenCalledTimes(1)
    expect(onFirstGetterRender).toHaveBeenCalledTimes(4)
    expect(onSecondGetterRender).toHaveBeenCalledTimes(4)
    expect(screen.getAllByTestId("getter")).toMatchSnapshot()
  })
})
