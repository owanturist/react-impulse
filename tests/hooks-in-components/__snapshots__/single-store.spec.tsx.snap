// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`single store Multiple Setters / Getters 1`] = `
Array [
  <span
    data-testid="getter"
  >
    0
  </span>,
  <span
    data-testid="getter"
  >
    0
  </span>,
]
`;

exports[`single store Multiple Setters / Getters 2`] = `
Array [
  <span
    data-testid="getter"
  >
    1
  </span>,
  <span
    data-testid="getter"
  >
    1
  </span>,
]
`;

exports[`single store Multiple Setters / Getters 3`] = `
Array [
  <span
    data-testid="getter"
  >
    2
  </span>,
  <span
    data-testid="getter"
  >
    2
  </span>,
]
`;

exports[`single store Multiple Setters / Getters 4`] = `
Array [
  <span
    data-testid="getter"
  >
    3
  </span>,
  <span
    data-testid="getter"
  >
    3
  </span>,
]
`;
